activerecord/lib/active_record/core.rb

```diff
@@ -824,7 +824,13 @@ def inspect_with_attributes(attributes_to_list)
       end
 
       def attributes_for_inspect
-        self.class.attributes_for_inspect == :all ? attribute_names : self.class.attributes_for_inspect
+        self.class.attributes_for_inspect == :all ? all_attributes_for_inspect : self.class.attributes_for_inspect
+      end
+
+      def all_attributes_for_inspect
+        return [] unless @attributes
+
+        attribute_names
```

Comment:
If we do `Model.allocate.inspect` with `attributes_for_inspect = :all` we will get a no method error since `@attributes` is `nil`. So I introduced this method. We can't do `self.class.attribute_names` since it will not include virtual columns.

Alternatively we could update active model to be resilient to `@attributes` being `nil`, but there are already `nil` checks for `@attributes` in this code so I thought it was okay to continue that practice. But I am open to feedback.

Reply:
We already do this check on line 812. 
Maybe `inspect_with_attributes` should not accept `attributes_to_list` but just call `attributes_for_inspect` itself?

Reply:
Yeah that's what I would do, however, we also need to pass this from `full_inspect` which also calls `inspect_with_attributes`.

Reply:
Maybe:
```ruby
      def inspect_with_attributes(attributes_for_inspect=self.class.attributes_for_inspect)
        inspection = if @attributes
          attributes_to_list = if attributes_for_inspect == :all
            all_attributes_for_inspect
          else
            self.class.attributes_for_inspect
          end
          attributes_to_list.filter_map do |name|
            name = name.to_s
            if _has_attribute?(name)
              "#{name}: #{attribute_for_inspect(name)}"
            end
          end.join(", ")
        else
          "not initialized"
        end
        "#<#{self.class} #{inspection}>"
      end

    def full_inspect
        inspect_with_attributes(:all)
    end
```

Reply:
The problem is that we can't call `self.class.attributes_for_inspect` because it won't include virtual columns (there is a test testing for this).

I did think about going in something like this direction but I am not sure if it's any less complicated 🤔

Reply:
I guess we could call `self.class.attributes_for_inspect` as the default param and then `self.attributes_for_inspect` in the method body but it's somewhat awkward.